src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,635 @@
+
+    private final ThreadContext threadContext;
+    private final ResourceSharingIndexHandler resourceSharingIndexHandler;
+    private final AdminDNs adminDNs;
+
+    public ResourceAccessHandler(
+        final ThreadPool threadPool,
+        final ResourceSharingIndexHandler resourceSharingIndexHandler,
+        AdminDNs adminDns
+    ) {
+        this.threadContext = threadPool.getThreadContext();
+        this.resourceSharingIndexHandler = resourceSharingIndexHandler;
+        this.adminDNs = adminDns;
+    }
+
+    /**
+     * Initializes the recipient types for users, roles, and backend roles.
+     * These recipient types are used to identify the types of recipients for resource sharing.
+     */
+    public void initializeRecipientTypes() {
+        RecipientTypeRegistry.registerRecipientType(Recipient.USERS.getName(), new RecipientType(Recipient.USERS.getName()));
+        RecipientTypeRegistry.registerRecipientType(Recipient.ROLES.getName(), new RecipientType(Recipient.ROLES.getName()));
+        RecipientTypeRegistry.registerRecipientType(
+            Recipient.BACKEND_ROLES.getName(),
+            new RecipientType(Recipient.BACKEND_ROLES.getName())
+        );
+    }
+
+    /**
+     *  Returns a set of accessible resource IDs for the current user within the specified resource index.
+     * @param resourceIndex The resource index to check for accessible resources.
+     * @param listener The listener to be notified with the set of accessible resource IDs.
+     *
+     */
+    public Future<Void> getAccessibleResourceIdsForCurrentUser(String resourceIndex, ActionListener<Set<String>> listener) {
+        final UserSubjectImpl userSubject = (UserSubjectImpl) threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        // If no user is authenticated, return an empty set
+        if (user == null) {
+            LOGGER.info("Unable to fetch user details.");
+            listener.onResponse(Collections.emptySet());
+            return null;
+        }
+
+        LOGGER.info("Listing accessible resources within the resource index {} for user: {}", resourceIndex, user.getName());
+
+        // 2. If the user is admin, simply fetch all resources
+        if (adminDNs.isAdmin(user)) {
+            loadAllResources(resourceIndex, ActionListener.wrap(listener::onResponse, listener::onFailure));
+            return null;
+        }
+
+        // StepListener for the user’s "own" resources
+        StepListener<Set<String>> ownResourcesListener = new StepListener<>();
+
+        // StepListener for resources shared with the user’s name
+        StepListener<Set<String>> userNameResourcesListener = new StepListener<>();
+
+        // StepListener for resources shared with the user’s roles
+        StepListener<Set<String>> rolesResourcesListener = new StepListener<>();
```

Comment:
TIL about the StepListener. Nice!

I see that its marked as [`@opensearch.internal`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/StepListener.java#L67), should we open up a PR in core to mark this public and solicit feedback?

Reply:
Looks like its used in quite a few places: https://github.com/search?q=org%3Aopensearch-project%20StepListener&type=code

I opened up a [PR in core](https://github.com/opensearch-project/OpenSearch/pull/17076) to mark this as `@opensearch.api`

Reply:
I saw the PR. Looks like we decided to not mark it as public.